### PR TITLE
Macro methods

### PR DESCRIPTION
**Pull request by bsideup**
Introduce macro methods and re-implement macro as a macro method

### CODE REVIEW

File: subprojects/groovy-macro/src/main/groovy/org/codehaus/groovy/macro/transform/MacroTransformation.java [hunk 1]
```diff
@@ -18,42 +18,113 @@
...
+
+                for (int i = 0; i < callArguments.size(); i++) {
+                    argumentsList[i] = ClassHelper.make(callArguments.get(i).getClass());
+                }
+
+                methods = StaticTypeCheckingSupport.chooseBestMethod(MACRO_CONTEXT_CLASS_NODE, methods, argumentsList);
+
+                for (MethodNode macroMethodNode : methods) {
+                    if (!(macroMethodNode instanceof ExtensionMethodNode)) {
+                        // TODO is it even possible?
```
Comment (melix) [inline subprojects/groovy-macro/src/main/groovy/org/codehaus/groovy/macro/transform/MacroTransformation.java]: If you don't know if it's possible, it's probably better to throw an error. Then if the case appears we know we need to fix it.

File: subprojects/groovy-macro/src/main/groovy/org/codehaus/groovy/macro/transform/MacroTransformation.java [hunk 2]
```diff
@@ -18,42 +18,113 @@
...
                 return sourceUnit;
             }
 
             @Override
-            public Expression transform(final Expression exp) {
-                if (exp instanceof ConstructorCallExpression) {
-                    MethodCallExpression call = exp.getNodeMetaData(MacroTransformation.class);
-                    if (call!=null) {
-                        return call;
+            public void visitMethodCallExpression(MethodCallExpression call) {
```
Comment (melix) [inline subprojects/groovy-macro/src/main/groovy/org/codehaus/groovy/macro/transform/MacroTransformation.java]: This method is very long. Could you split it into smaller focused pieces?

File: subprojects/groovy-macro/src/main/groovy/org/codehaus/groovy/macro/transform/MacroMethodsCache.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+import java.util.*;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * TODO share some code with {@link org.codehaus.groovy.transform.stc.StaticTypeCheckingSupport.ExtensionMethodCache}
+ * @author Sergei Egorov <bsideup@gmail.com>
+ * @since 2.5.0
+ */
+class MacroMethodsCache {
```
Comment (danielsun1106) [inline subprojects/groovy-macro/src/main/groovy/org/codehaus/groovy/macro/transform/MacroMethodsCache.java]: Maybe you can reuse LRU cache in the parrot branch: https://issues.apache.org/jira/browse/GROOVY-7977

Comment (bsideup) [inline subprojects/groovy-macro/src/main/groovy/org/codehaus/groovy/macro/transform/MacroMethodsCache.java]: We could once parrot is merged to master :)

**Comment by paulk-asert** [comment]
Proposed PR merged. Thanks Sergei!

**Pull request closed without merging**